### PR TITLE
Klipper 12864 improve

### LINKED ISSUE #290: [FR] [ST7920_Simulator.c] Add a character generator in order to use klipper
Many users of your skr boards use klipper (https://github.com/KevinOConnor/klipper).

Please add a character generator to TFT / src / User / API / UI / ST7920_Simulator.c to make your tft24 screen and similiar usable.

Further information can be found in the Klipper issue at https://github.com/KevinOConnor/klipper/issues/2231

Your community will appreciate it very much

Thank you very much

### PR DESCRIPTION
**Pull request by Msq001**
### Requirements

https://github.com/KevinOConnor/klipper/pull/3979 This PR of Klipper has solved the SPI data shift problem of TFT 12864 mode(Thanks @teeminus)
According to the setting in https://github.com/KevinOConnor/klipper/pull/3979#issuecomment-784215162, the TFT 12864 can work normally with Klipper

We need to set `lcd_type` to `emulated_st7920`, and set the correct pin of `software_spi`

```
[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
```

### Description

TFT work with SKR-2 | Marlin & Full Screen OFF | Marlin & Full Screen ON | Klipper & Full Screen OFF | Klipper & Full Screen ON
-- | -- | -- | -- | --
TFT24 V1.1 | 💚good  | ⚠️beta | 💚good | ⚠️beta
TFT35 E3 V3.0 | 💚good  | 💚good  | 💚good | 💚good
TFT50 V3.0 | 💚good  | 💚good  |  💚good| 💚good
TFT70 V3.0 | 💚good  |  💚good | 💚good | 💚good

It reduces ram loss of hundreds of bytes and increases flash of tens of bytes
### Related Issues
fixed #290

### CODE REVIEW

**Comment by oldman4U** [comment]
Hi.

Should we ask the Klipper users to test this PR?

**Comment by Msq001** [comment]
This feature can be used normally, but the code needs more cleaning. After cleaning, I will convert to `Ready for review`, and then ask members to test

**Comment by oldman4U** [comment]
Perfect. 

Thank you

File: TFT/src/User/API/boot.c [hunk 1]
```diff
@@ -338,6 +338,7 @@ void scanUpdates(void)
       if (updateFont(FONT_ROOT_DIR "/byte_ascii.fon", BYTE_ASCII_ADDR) &&
           updateFont(FONT_ROOT_DIR "/word_unicode.fon", WORD_UNICODE) &&
           updateFont(FONT_ROOT_DIR "/large_byte_ascii.fon", LARGE_FONT_ADDR) &&
+          updateFont(FONT_ROOT_DIR "/16x8_byte_ascii.fon", _16X8_FONT_ADDR) &&
```
Comment (guruathwal) [inline TFT/src/User/API/boot.c:341]: @Msq001 Please rename 16x8 to 8x16

File: TFT/src/User/API/UI/ST7920_Emulator.c [hunk 2]
```diff
@@ -139,26 +165,26 @@ void ST7920_DispHCGROM(uint8_t p)
...
   getCharacterInfo(&p, &info);
-
-  uint8_t  i = 0;
-  uint8_t  font[BYTE_HEIGHT * BYTE_WIDTH / 8];
+  uint16_t bitMapSize =(info.pixelHeight * info.pixelWidth / 8);
+  uint8_t  font[bitMapSize];
   uint32_t temp = 0;
 
-  W25Qxx_ReadBuffer(font, info.bitMapAddr, BYTE_HEIGHT * BYTE_WIDTH / 8);
+    W25Qxx_ReadBuffer(font, info.bitMapAddr, bitMapSize);
```
Comment (digant73) [inline TFT/src/User/API/UI/ST7920_Emulator.c:177]: too much indentation

File: TFT/src/User/API/UI/ST7920_Emulator.h [hunk 3]
```diff
@@ -238,11 +238,27 @@ typedef struct
...
+  ST7920_REG reg;
+  // Character Generation RAM
+  uint8_t CGRAM[64][2];  // [64*2] = [4 * 16*2*8], means 4 * [16*16] bitmap font
+  // Display Data RAM
+  uint8_t DDRAM[4][16];  // 16 characters (8x16) by 4 lines
+  // Graphic Display RAM
+  uint8_t GDRAM[64][16];  // [64*16] = [64 * 16*8], means 64 * 128 pixels
+  // 8x16 ASCII font
+  uint8_t _8x16Font[95][16];  // 0x20 ~ 0x7E = 95char, 8width * 16height / 8 = 16 bytes, scan dir: left to right & top to bottom
+}ST7920;
```
Comment (digant73) [inline TFT/src/User/API/UI/ST7920_Emulator.h:259]: a space after }

**Review by digant73 — CHANGES_REQUESTED** [review]

**Comment by oldman4U** [comment]
So the reaction of the community is overwhelming....

Have you got something back?

**Comment by Talha909** [comment]
My lcd show the result on screen. But rotary switch did not work.  What should i do

**Comment by jasc76** [comment]
Tested this on TFT24 with MKS GEN L with the provided [display] section above. 

Klippy Error:
Unknown value 'EXP2_1' in enumeration 'pin'
This type of error is frequently caused by running an older
version of the firmware on the micro-controller (fix by
recompiling and flashing the firmware).

Known versions: host=v0.9.1-533-g27f8cf02, mcu=v0.9.1-528-gc0d86048

printer.cfg
aliases:
    # EXP1 header
    EXP1_1=ar37, EXP1_3=ar17, EXP1_5=ar23, EXP1_7=ar27, EXP1_9=<GND>,
    EXP1_2=ar35, EXP1_4=ar16, EXP1_6=ar25,  EXP1_8=ar29, EXP1_10=<5V>

    # EXP2 header
    EXP2_1=ar50, EXP2_3=ar31, EXP2_5=ar33, EXP2_7=ar49, EXP2_9=<GND>,
    EXP2_2=ar52, EXP2_4=ar53, EXP2_6=ar51,  EXP2_8=<RST>, EXP2_10=ar41

[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2
[output_pin beeper]
pin: EXP1_1

**Comment by bigtreetech** [comment]
> My lcd show the result on screen. But rotary switch did not work. What should i do

Which motherboard do you use? Maybe there is something wrong with the setting of 
```
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
```

**Comment by bigtreetech** [comment]
> Klippy Error:
> Unknown value 'EXP2_1' in enumeration 'pin'
> This type of error is frequently caused by running an older
> version of the firmware on the micro-controller (fix by
> recompiling and flashing the firmware).
> 
> Known versions: host=v0.9.1-533-g27f8cf02, mcu=v0.9.1-528-gc0d86048

Do you using the latest version of klipper and the MCU of MKS GEN L has also updated the latest version of firmware.bin?

**Comment by jasc76** [comment]
Known versions: host=v0.9.1-533-g27f8cf02, mcu=v0.9.1-533-g27f8cf02, rpi=v0.9.1-533-g27f8cf02

no change, error still is there. Its the latest kiauh is offering

What I discovered:
Changing
spi_software_miso_pin: EXP2_1
to
spi_software_miso_pin: EXP2_**2** (was just an idea, what happens) 
the error goes away and it wors

**Comment by Talha909** [comment]
I am using skr 1.4 Tubo. I got the display on the screen but my rotary switch did not work and there is no beeper. What should i do

**Comment by oldman4U** [comment]
Contact your dealer.

**Comment by bigtreetech** [comment]
> I am using skr 1.4 Tubo. I got the display on the screen but my rotary switch did not work and there is no beeper. What should i do

Hello, Can you confirm that the following pins are set properly? If you are still unable to work after confirmation, please provide your `printer.cfg` file
```
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2

[output_pin beeper]
pin: EXP1_1
```

**Comment by Talha909** [comment]
HI, Still facing the issue. My rotary switch is not working. I am using SKR 1.4 Tubo, TMC2130 and TFT 24
[printer (5).zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/6621190/printer.5.zip)

**Comment by gregmatic** [comment]
my screen shows Marlin Mode with blank screen. can anyone confirm the following works with SKR E3 Mini v1.2 & TFT35 E3 V3.0?

[board_pins]
aliases:
    # EXP1 header
    EXP1_1=PB5, EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8, EXP1_9=<GND>,
    EXP1_2=PB6, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB7, EXP1_10=<5V>

[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2

**Comment by bigtreetech** [comment]
> my screen shows Marlin Mode with blank screen. can anyone confirm the following works with SKR E3 Mini v1.2 & TFT35 E3 V3.0?
> 
> [board_pins]
> aliases:
> # EXP1 header
> EXP1_1=PB5, EXP1_3=PA9, EXP1_5=PA10, EXP1_7=PB8, EXP1_9=,
> EXP1_2=PB6, EXP1_4=, EXP1_6=PB9, EXP1_8=PB7, EXP1_10=<5V>
> 
> [display]
> lcd_type: emulated_st7920
> spi_software_miso_pin: EXP2_1
> spi_software_mosi_pin: EXP1_3
> spi_software_sclk_pin: EXP1_5
> en_pin: EXP1_4
> encoder_pins: ^EXP2_3, ^EXP2_5
> click_pin: ^!EXP1_2

According to the instructions here
https://github.com/KevinOConnor/klipper/blob/478f26cab6c8e73e3970d2b303bc818f6910ae74/config/sample-lcd.cfg#L49
SKR-mini-E3-V1.2 has no EXP1 + EXP2 and only the Ender-3 stock display port, so the settings should be as follows


```
[display]
lcd_type: st7920
cs_pin: EXP1_7
sclk_pin: EXP1_6
sid_pin: EXP1_8
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2

[output_pin beeper]
pin: EXP1_1
```

**Comment by gregmatic** [comment]
> [display]
lcd_type: st7920
cs_pin: EXP1_7
sclk_pin: EXP1_6
sid_pin: EXP1_8
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2

[output_pin beeper]
pin: EXP1_1


thank you...i tried this but Marlin Mode still shows black screen. I didn't move any wiring ...the TFT35 worked fine in both Marlin Mode & Touch Mode when my firmware was Marlin...

**Comment by Terixxx** [comment]
How it is possible to connect TFT3.5 V3.0 to BTT Octopus V1.1 
[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
Тhese pins are totally not as in board pin mapping. And how to be sure which pinout is correct. The official config file in repository for Octopus board is: 

[board_pins]
aliases:
    # EXP1 header
    EXP1_1=PE8, EXP1_2=PE7,
    EXP1_3=PE9, EXP1_4=PE10,
    EXP1_5=PE12, EXP1_6=PE13,    # Slot in the socket on this side
    EXP1_7=PE14, EXP1_8=PE15,
    EXP1_9=<GND>, EXP1_10=<5V>,

    # EXP2 header
    EXP2_1=PA6, EXP2_2=PA5,
    EXP2_3=PB1, EXP2_4=PA4,
    EXP2_5=PB2, EXP2_6=PA7,      # Slot in the socket on this side
    EXP2_7=PC15, EXP2_8=<RST>,
    EXP2_9=<GND>, EXP2_10=<5V>

But how this is possible when:
**spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5**

but regarding on https://raw.githubusercontent.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/master/Hardware/BIGTREETECH%20Octopus.pdf
SPI1_MISO -> PA6
SPI1_MOSI -> PA7
SPI1_SCK    -> PA5
FORGET IT!!!!!!!! I am totally lost in this wrong diagrams and pinouts!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
https://github.com/bigtreetech/BIGTREETECH-OCTOPUS-V1.0/blob/master/Octopus%20works%20on%20Voron%20v2.4/Firmware/Klipper/BTT_OctoPus_Voron2_Config.cfg
I can't believe that this is happens with BTT OOCTOPUS BOARD!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! This is not possible.

**Comment by Terixxx** [comment]
> 
> ### Requirements
> 
> [KevinOConnor/klipper#3979](https://github.com/KevinOConnor/klipper/pull/3979) This PR of Klipper has solved the SPI data shift problem of TFT 12864 mode(Thanks @teeminus)
> According to the setting in [KevinOConnor/klipper#3979 (comment)](https://github.com/KevinOConnor/klipper/pull/3979#issuecomment-784215162), the TFT 12864 can work normally with Klipper
> 
> We need to set `lcd_type` to `emulated_st7920`, and set the correct pin of `software_spi`
> 
> ```
> [display]
> lcd_type: emulated_st7920
> spi_software_miso_pin: EXP2_1
> spi_software_mosi_pin: EXP1_3
> spi_software_sclk_pin: EXP1_5
> en_pin: EXP1_4
> encoder_pins: ^EXP2_3, ^EXP2_5
> click_pin: ^!EXP1_2
> ```
> 
> ### Description
> TFT work with SKR-2 	Marlin & Full Screen OFF 	Marlin & Full Screen ON 	Klipper & Full Screen OFF 	Klipper & Full Screen ON
> TFT24 V1.1 	💚good 	⚠️beta 	💚good 	⚠️beta
> TFT35 E3 V3.0 	💚good 	💚good 	💚good 	💚good
> TFT50 V3.0 	💚good 	💚good 	💚good 	💚good
> TFT70 V3.0 	💚good 	💚good 	💚good 	💚good
> 
> It reduces ram loss of hundreds of bytes and increases flash of tens of bytes
> ### Benefits
> ### Related Issues
> 
> fixed #290

Could you please provide Pin map and board corresponding to this configuration

**Comment by hapklaar** [comment]
@Msq001 This is working for me, except for that the screen gets garbled/scrambled after a few minutes of printing. Tried both ways of addressing the screen, doesn't seem to matter for this specific problem.

```
[display]
lcd_type: st7920 
cs_pin: EXP1_4
sclk_pin: EXP1_5
sid_pin: EXP1_3
encoder_pins: ^EXP2_5, ^EXP2_3
click_pin: ^!EXP1_2
```

and
```
[display]
lcd_type: emulated_st7920
spi_software_miso_pin: EXP2_1
spi_software_mosi_pin: EXP1_3
spi_software_sclk_pin: EXP1_5
en_pin: EXP1_4
encoder_pins: ^EXP2_3, ^EXP2_5
click_pin: ^!EXP1_2
```

**Comment by vMeph** [comment]
@Msq001 could you provide the config for the btt SKR2 to testout
thanks in advanced

**Comment by vMeph** [comment]
its working !!!

**Comment by mlaci001** [comment]
The same is true with the E3 turbo and TFT35-E3-v3.
The screen is blank.
If the TFT button is held down for 3 seconds, the mode selection window will appear, here you have to select the Marlin mode (no matter what mode the screen is in).
The screen will still be blank, but if you press the button again, the menu will appear and work from there.
You have to do it again every time you start.
I don't know why it's not good at startup.
I can't use emulated_st7920 mode, the motherboard only has an EXP1 connector.

**Comment by chaicracker** [comment]
Hello everyone :)

Currently trying to get a BTT TFT35 V3 (the wide one, not the E3 version) working with an SKR Mini E3 V2.0

this is the configuration I could muster so far, though none of the tries suceeded so far. I had installed the fitting NoTouchScreenFirmware for the Display and says "ST7920Emulated ready"

`[board_pins]
aliases:
    EXP1_1=PB5,  EXP1_3=PA9,   EXP1_5=PA10, EXP1_7=PB8,  EXP1_9=<GND>,
    EXP1_2=PA15, EXP1_4=<RST>, EXP1_6=PB9,  EXP1_8=PB15, EXP1_10=<5V>,


[display]
lcd_type: emulated_st7920
en_pin: EXP1_7
spi_software_sclk_pin: EXP1_6
spi_software_mosi_pin: EXP1_8
spi_software_miso_pin: PA3
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2`

Cheers :)

**Comment by douglasdtlltd** [comment]
> 

This works with the current compile of the BTT firmware and a BTT SKR Mini E3 V2 with THESE PINS.

IM SO HAPPY I FOUND THIS.

**Pull request merged**